### PR TITLE
[BUG] Fix UNSUBs after disconnect can cause auth violations

### LINKED ISSUE #1320: NATS Client "Authorization Violation" after server restart
### Observed behavior

During reconnect, UNSUB messages  may be enqueued in the reconnectOutgoing queue if they are issued after the client’s connection status check but before the transition into reconnect mode. Under a race condition, UNSUB commands can be batched and transmitted prior to the mandatory CONNECT message. This causes the server to interpret the messages as coming from an unauthorized, not-yet-authenticated client, leading to Authorization Violation errors.

### Expected behavior

Reconnect is successful

### Server and client version

- NATS Server version: 2.10.24
- NATS Client library/version: 2.21.0
- Authentication method: Token Authentication
- Deployment: Multiple NATS server instances

### Host environment

_No response_

### Steps to reproduce

The analysis identified that if context switching occurs, threads performing unsubscribe operations might pass a connection check that becomes outdated once the reconnect process starts. As a result, UNSUB messages are sent with the wrong ordering relative to CONNECT, and the server responds with errors. The logs show that multiple errors can occur if several messages are sent in this state.

| T1 (NatsConnection.unsubscribe)                                                                                                                                 | T2 NatsConnection.closeSocket()                                                                                                                              |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Image                                                                       | Image                                                                    |
| Image The status is being checked, which will change in the next step in T2 |                                                                                                                                                              |
|                                                                                                                                                                 | Image                                                                    |
|                                                                                                                                                                 | Image                                                                    |
|                                                                                                                                                                 | Image writer.setReconnectMode() toggles the NatsConnectionWriter queue |
| Image                                                                       |                                                                                                                                                              |
| Image go to if branch                                               |                                                                                                                                                              |

I suppose the internal queue should be utilized solely by the driver’s internal logic. While subscription messages are correctly separate by client and internal flow, unsubscription commands are currently enqueued by client code, which violates the intended design. It is preferable that the internal queue be populated only by the driver; client code should not have the ability to directly enqueue internal messages.

Additionally, the current implementation only removes the authentication error for the specific server instance that successfully reestablishes a connection, leaving entries in the serverAuthErrors map for other instances. As a result, if a client later reconnects to a server that previously returned an authentication error, the stale error in serverAuthErrors may trigger unexpected authentication issues and close connection even after a successful connection on the other node.

### PR DESCRIPTION
**Pull request by scottf**
Fixes #1320

An Authorization Violation could occur if an UNSUB was sent between a disconnect and reconnect, because the UNSUB got queued and sent before the CONNECT protocol finished. This PR prevents that from happening and also addresses not needing to send unsubs after a disconnect.

1. Puts UNSUBS in the normal (not internal/reconnect) queue. This by itself is enough to solve the bug, since the normal queue is never sent while not connected.

2. On a disconnect, PINGS and PONGS were already filtered out of the normal outgoing queue because they are invalid anyway. 
    * This PR filters out UNSUB messages, since they are unnecessary traffic; subs are closed on the server anyway because of the disconnect.
    * This PR filters out SUB messages. By the time the request is made to send the actual SUB message to the server, the sub has been registered internally. This means that if there is a disconnect, the client is going to resub, so not filtering would actually create a double sub. This is a rare edge case. 

---
If you want to verify, run the test it passes. Then go into NatsConnection and change line 1062 to
```
queueInternalOutgoing(new ProtocolMessage(bab));
```
and then run the test, it will fail.

---
Thanks @ajax-surovskyi-y

### CODE REVIEW

File: src/main/java/io/nats/client/impl/NatsConnection.java [hunk 1]
```diff
@@ -1527,7 +1527,7 @@ public Duration RTT() throws IOException {
         pongQueue.add(pongFuture);
         try {
             long time = NatsSystemClock.nanoTime();
-            writer.queueInternalMessage(new ProtocolMessage(OP_PING_BYTES));
+            writer.queueInternalMessage(new ProtocolMessage(PONG_PROTO));
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/NatsConnection.java]: Use the premade PONG_PROTO for faster construction

File: src/main/java/io/nats/client/impl/NatsConnectionWriter.java [hunk 2]
```diff
@@ -105,10 +106,7 @@ Future<Boolean> stop() {
             try {
                 this.outgoing.pause();
                 this.reconnectOutgoing.pause();
-                // Clear old ping/pong requests
-                this.outgoing.filter((msg) ->
-                    msg.isProtocol() &&
-                        (msg.getProtocolBab().equals(OP_PING_BYTES) || msg.getProtocolBab().equals(OP_PONG_BYTES)));
+                this.outgoing.filter(NatsMessage::isProtocolFilterOnStop);
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/NatsConnectionWriter.java:109]: Leverage the new flag for faster filtering instead of doing byte compares on proto bytes

File: src/main/java/io/nats/client/impl/NatsConnectionWriter.java [hunk 3]
```diff
@@ -151,7 +149,7 @@ void sendMessageBatch(NatsMessage msg, DataPort dataPort, StatisticsCollector st
                 sendBuffer[sendPosition++] = CR;
                 sendBuffer[sendPosition++] = LF;
 
-                if (!msg.isProtocol()) { // because a protocol message does not have headers
+                if (!msg.isProtocol()) { // because a protocol message does not have headers or data
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/NatsConnectionWriter.java:152]: This clarification was important.

File: src/main/java/io/nats/client/impl/ProtocolMessage.java [hunk 4]
```diff
@@ -19,30 +19,39 @@
 // Protocol message is a special version of a NatsPublishableMessage extends NatsMessage
 // ----------------------------------------------------------------------------------------------------
 class ProtocolMessage extends NatsPublishableMessage {
+    final boolean filterOnStop;
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/ProtocolMessage.java:22]: must construct with this information now

File: src/main/java/io/nats/client/impl/ProtocolMessage.java [hunk 5]
```diff
@@ -19,30 +19,39 @@
...
         protocolBab = babProtocol;
         sizeInBytes = controlLineLength = protocolBab.length() + 2; // CRLF, protocol doesn't have data
+        this.filterOnStop = filterOnStop;
     }
 
     ProtocolMessage(byte[] protocol) {
         super(false);
         protocolBab = new ByteArrayBuilder(protocol);
         sizeInBytes = controlLineLength = protocolBab.length() + 2; // CRLF, protocol doesn't have data
+        this.filterOnStop = true;
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/ProtocolMessage.java:35]: I looked through every use this constructor. 99% were test construction. The only production usage were these, the pre-made messages which speed up new ProtocolMessage creation by sharing the existing protocol buffer which never changes for these fixed size protocols. These are used to construct new ProtocolMessage, which is required since they have some state related to their queueing - you can only share / reuse the data, not the message itself.
```
private static final ProtocolMessage PING_PROTO = new ProtocolMessage(OP_PING_BYTES);
private static final ProtocolMessage PONG_PROTO = new ProtocolMessage(OP_PONG_BYTES);
```

**Review by MauriceVanVeen — APPROVED** [review]

File: src/main/java/io/nats/client/impl/ProtocolMessage.java (hunk 4 above)
Comment (ajax-surovskyi-y) [reply src/main/java/io/nats/client/impl/ProtocolMessage.java:22]: One concern I wanted to raise is the potential memory impact of adding additional state to each NatsMessage.
In high-throughput systems (we’re handling tens of thousands of messages per second), even a small increase in object size can become noticeable.

Comment (ajax-surovskyi-y) [reply src/main/java/io/nats/client/impl/ProtocolMessage.java:22]: On second thought, I may have overestimated the impact — this change applies to protocol messages rather than the actual PUB data, so the volume is relatively low.
That likely means the memory overhead is negligible.
